fix: clear omit_containers, additional_fqdns, additional_hostnames config with empty flag, fixes #6140

jonesrussell opened the pull request:
## Issue

* `ddev config --omit-containers=""` doesn't remove `omit_containers` from `.ddev/config.yaml` as expected (#6140).
* `ddev config --additional-hostnames` and `ddev config --additional-fqdns` have the same problem.

## Solution

This PR clears `omit-containers` when an empty string is passed. It uses the `Changed` method, consistent with the global version of the command. Changes are made in `handleMainConfigArgs` in `cmd/ddev/cmd/config.go`.

## Testing

Manual testing involves running `ddev config --omit-containers=db` and `ddev config --omit-containers=""` and checking `.ddev/config.yaml`. Automated tests will be added to verify the flag's functionality.

## Related Issues

Fixes issue #6140.

## Deployment

No special deployment actions needed. This change doesn't affect other code.

rfay commented:
Thanks for this! Could you please evaluate the `ddev config global --omit-containers` situation also, and poke around just a bit to see if we have other commands that deal with arrays that aren't clear-able?

rfay commented:
Rebased, looking forward to your work on this.

rfay commented:
Hoping you can take a look, it would be good to get these things taken care of together. Thanks!

rfay commented:
Rebased.

stasadev commented:
Another flag with the same problem: `additional_fqdns`

spoonerWeb commented:
> Another flag with the same problem: `additional_fqdns`

And `additional_hostnames`

jonesrussell commented:
The global flag is already removing omit_containers as expected.

I added checks for additional-hostnames and additional-fqdns.

There was code duplication, so I added a `processFlag` function.

rfay commented:
You're awesome, thanks!

rfay commented:
Pushed a commit to pacify golangci-lint, the `else` was redundant there, https://github.com/ddev/ddev/pull/6202/commits/d0ad534e6d51a1f9f4d8ba795624dc099ec7095a

You can precheck this with `make staticrequired`

jonesrussell commented:
Ah thanks for the linter check heads up.

After looking at processFlag again I thought it reads better if we return early when flag is not changed.

rfay commented:
I started looking at this and found that there were a few other array-type arguments, then added test coverage and found that we were failing. There are some things that definitely work under bash (where quotes get processed automatically) and don't work in some other contexts.

* Added other items
* Added project test coverage
* Added global items
* Added global test coverage

And of course I discovered that we should have been using the `StringSlice` type in Cobra (like we do for `upload-dirs`) for all of these, and our life would have been so much easier. I tried converting some of these to StringSlice, but it ended up being a little risky and likely to change the existing behavior.

https://github.com/ddev/ddev/blob/f6977639ad2ee95667352ccb527371881abe8954/cmd/ddev/cmd/config.go#L233

Now it needs a whole new review and manual testing, including testing of all the affected array types.

stasadev reviewed (approved):
Looks good to me.

stasadev commented on cmd/ddev/cmd/config_test.go:
```diff
@@ -284,21 +276,33 @@ func TestConfigSetValues(t *testing.T) {
 	assert.Equal(app.ComposerRoot, "")
 	assert.Equal(app.WebImage, "")
 	assert.Equal(len(app.WorkingDir), 0)
+	assert.Empty(app.AdditionalHostnames)
+	assert.Empty(app.AdditionalFQDNs)
+	assert.Empty(app.OmitContainers)
+	assert.Empty(app.UploadDirs)
+	assert.Empty(app.WebEnvironment)
```
```suggestion
	assert.Empty(app.AdditionalHostnames)
	assert.Empty(app.AdditionalFQDNs)
	assert.Empty(app.DBImageExtraPackages)
	assert.Empty(app.OmitContainers)
	assert.Empty(app.UploadDirs)
	assert.Empty(app.WebEnvironment)
	assert.Empty(app.WebImageExtraPackages)
```

stasadev commented on cmd/ddev/cmd/config_test.go:
```diff
@@ -129,34 +129,30 @@ func TestConfigWithSitenameFlagDetectsDocroot(t *testing.T) {
 		err = os.Chdir(origDir)
 		assert.NoError(err)
-		out, err := exec.RunHostCommand(DdevBin, "delete", "-Oy", t.Name())
+		out, err := exec.RunHostCommand(DdevBin, "delete", "-Oy", strings.ToLower(t.Name()))
```
```suggestion
		out, err := exec.RunHostCommand(DdevBin, "delete", "-Oy", projectName)
```

stasadev commented on docs/content/users/usage/commands.md:
```diff
@@ -189,8 +189,8 @@ ddev config --webserver-type=apache-fpm
 
-* `--additional-fqdns`: Comma-delimited list of project FQDNs.
-* `--additional-hostnames`: Comma-delimited list of project hostnames.
+* `--additional-fqdns`: Comma-delimited list of project FQDNs or `--additional-fqdns=""` to remoe any configured FQDNs.
```
```suggestion
* `--additional-fqdns`: Comma-delimited list of project FQDNs or `--additional-fqdns=""` to remove any configured FQDNs.
```

stasadev commented on docs/content/users/usage/commands.md:
```diff
@@ -261,7 +261,7 @@ ddev config global --omit-containers=ddev-ssh-agent
 * `--mailpit-https-port`: The Mailpit HTTPS port *default* for all projects; can be overridden by project configuration.
 * `--no-bind-mounts`: If `true`, don’t use bind-mounts. Useful for environments like remote Docker where bind-mounts are impossible. (default is equal to `--no-bind-mounts=true`)
-* `--omit-containers`: For example, `--omit-containers=ddev-ssh-agent`.
+* `--omit-containers`: For example, `--omit-containers=ddev-ssh-agent` or `--omit-containers=""".
```
```suggestion
* `--omit-containers`: For example, `--omit-containers=ddev-ssh-agent` or `--omit-containers=""`.
```

rfay commented:
Thanks @jonesrussell !
